cmd/gitcloner/gogit/cloner.go

```diff
@@ -0,0 +1,174 @@
+package gogit
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	httpgit "github.com/go-git/go-git/v5/plumbing/transport/http"
+	gossh "github.com/go-git/go-git/v5/plumbing/transport/ssh"
+	"github.com/rancher/gitjob/cmd/gitcloner/cmd"
+	giturls "github.com/whilp/git-urls"
+	"golang.org/x/crypto/ssh"
+)
+
+const defaultBranch = "master"
+
+var (
+	plainClone = git.PlainClone
+	readFile   = os.ReadFile
+)
+
+type Cloner struct{}
+
+type Options struct {
+	Repo            string
+	Branch          string
+	Auth            transport.AuthMethod
+	InsecureSkipTLS bool
+	CABundle        []byte
+}
+
+func NewCloner() *Cloner {
+	return &Cloner{}
+}
+
+func (c *Cloner) CloneRepo(opts *cmd.Options) error {
+	auth, err := createAuthFromOpts(opts)
+	if err != nil {
+		return err
+	}
+	caBundle, err := getCABundleFromFile(opts.CABundleFile)
+	if err != nil {
+		return err
+	}
+
+	if opts.Branch == "" && opts.Revison == "" {
+		opts.Branch = defaultBranch
+		return cloneBranch(opts, auth, caBundle)
+	}
+
+	if opts.Branch != "" {
+		return cloneBranch(opts, auth, caBundle)
+	}
+
+	return cloneRevision(opts, auth, caBundle)
+}
+
+func cloneBranch(opts *cmd.Options, auth transport.AuthMethod, caBundle []byte) error {
+	_, err := plainClone(opts.Path, false, &git.CloneOptions{
+		URL:             opts.Repo,
+		Auth:            auth,
+		InsecureSkipTLS: opts.InsecureSkipTLS,
+		CABundle:        caBundle,
+		SingleBranch:    true,
+		ReferenceName:   plumbing.ReferenceName(opts.Branch),
+	})
+
+	return err
+}
+
+func cloneRevision(opts *cmd.Options, auth transport.AuthMethod, caBundle []byte) error {
+	r, err := plainClone(opts.Path, false, &git.CloneOptions{
+		URL:             opts.Repo,
+		Auth:            auth,
+		InsecureSkipTLS: opts.InsecureSkipTLS,
+		CABundle:        caBundle,
+	})
+	if err != nil {
+		return err
+	}
+	h, err := r.ResolveRevision(plumbing.Revision(opts.Revison))
+	if err != nil {
+		return err
+	}
+	w, err := r.Worktree()
+	if err != nil {
+		return err
+	}
+
+	return w.Checkout(&git.CheckoutOptions{
+		Hash: *h,
+	})
+}
+
+func getCABundleFromFile(path string) ([]byte, error) {
+	if path != "" {
+		caBundle, err := readFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		return caBundle, nil
+	}
+
+	return nil, nil
```

Comment:
nit-pick: early returns are also preferred over big if blocks:
```suggestion
	if path == "" {
		return nil, nil
	}
	return readFile(path)
}
```